File: python/microservice_toolbox/config/loader.py
#!/usr/bin/env python
# coding:utf-8

"""
ESSENTIAL PROCESS:
Initializes a configuration loader following the Microservice Toolbox 'Hierarchy of Truth'.
Ensures service settings remain consistent across standalone and production fleets.

DATA FLOW:
1. Load base configuration from YAML profile.
2. Apply context-aware overrides (standalone/test modes).
3. Merge CLI overrides (highest priority).

KEY PARAMETERS:
- profile: The active configuration profile (e.g., standalone, production).
- specific_flags: Optional list of CLI flags to parse.
"""

from json import dumps as jsonDumps
from json import loads as jsonLoads
from os import getenv as osGetenv
from os.path import exists as osPathExists
from typing import Any, Callable, Dict, List, Optional

from yaml import safe_load as yamlSafe_load

from ..utils.logger import ILogger, ensure_safe_logger
from .args import parse_cli_args
from .lib_loader import CALLBACK_TYPE, lib
from .merger import deep_merge

# -----------------------------------------------------------------------------------------------




def load_config(
    profile: str, specific_flags: Optional[List[str]] = None, input_args: Optional[List[str]] = None
) -> "AppConfig":
    """Semantic helper to match Go LoadConfig()."""
    return AppConfig(profile, specific_flags, input_args=input_args)


# -----------------------------------------------------------------------------------------------


def load_config_with_logger(
    profile: str, logger: Optional[ILogger], specific_flags: Optional[List[str]] = None
) -> "AppConfig":
    """Semantic helper to match Go LoadConfigWithLogger()."""
    return AppConfig(profile, specific_flags, logger=logger)


# -----------------------------------------------------------------------------------------------


class AppConfig:
    """
    AppConfig wraps configuration data and provides layered resolution logic.
    It ensures that service settings remain consistent whether running in
    standalone development mode or across a containerized production fleet.
    """

    Name = "AppConfig"

    # -----------------------------------------------------------------------------------------------

    def __init__(
        self,
        profile: str,
        specific_flags: Optional[List[str]] = None,
        logger: Optional[ILogger] = None,
        input_args: Optional[List[str]] = None,
    ):
        self.profile = profile
        self.data: Dict[str, Any] = {}
        self.logger = ensure_safe_logger(logger)
        self.cli_args = parse_cli_args(specific_flags, input_args=input_args)
        filename = f"{profile}.yaml"

        # ---------------------------------------------------------------------
        # BRIDGE INITIALIZATION (v1.9.8 Standard)
        # ---------------------------------------------------------------------
        self._handle = None
        self._callback_refs = {} # Keep references to avoid GC

        if lib:
            try:
                self._handle = lib.DistConf_New(profile.encode('utf-8'))
                if self._handle:
                    self.logger.info("{0} : libdistconf session initialized (handle: {1})".format(self.Name, self._handle))
                    # Sync local data with bridge state
                    self._sync_from_bridge()
            except Exception as e:
                self.logger.warning("{0} : libdistconf initialization failed: {1}. Falling back to native.".format(self.Name, e))

        # ---------------------------------------------------------------------
        # PHASE 1: Load base configuration from YAML (Native Fallback)
        # ---------------------------------------------------------------------
        if not self._handle:
            if not osPathExists(filename):
                # Try fallback to config/ folder
                filename = f"config/{profile}.yaml"
                if not osPathExists(filename):
                    raise FileNotFoundError(f"Toolbox (Python): Config file not found for profile '{profile}' (Checked {profile}.yaml and config/{profile}.yaml)")
            self._load_from_file(filename)

        # ---------------------------------------------------------------------
        # PHASE 2: Apply context-aware overrides
        # In standalone/test modes, we ensure the local file is authoritative
        # over any previous merges (matching Go behavior).
        # ---------------------------------------------------------------------
        is_dev = profile in ["standalone", "test"]
        if is_dev:
            self.logger.info("{0} : Dev Mode detected. Re-applying Local File as Hard Override.".format(self.Name))
            self._apply_file_override(filename)
        else:
            self.logger.info("{0} : Production Mode detected. Configuration state remains stable.".format(self.Name))

        # ---------------------------------------------------------------------
        # PHASE 3: Apply CLI Overrides (The absolute Highest Priority)
        # ---------------------------------------------------------------------
        self._apply_cli_overrides()

        # If --key flag provided, set it as ENV override for the Private Key (decryption engine)
        if self.cli_args.key:
            from os import environ as os_environ
            os_environ["BASTIEN_PRIVATE_KEY_PATH"] = self.cli_args.key

        # ---------------------------------------------------------------------
        # PHASE 4: Load Public Key
        # ---------------------------------------------------------------------
        self._load_public_key()

    # -----------------------------------------------------------------------------------------------

    def _load_public_key(self) -> None:
        """Finds and loads public.pem into self.data['common']['public_key']."""
        path = osGetenv("BASTIEN_PUBLIC_KEY_PATH")
        if not path:
            path = "/etc/bastien/public.pem"
            if not osPathExists(path):
                if osPathExists("./public.pem"):
                    path = "./public.pem"
                else:
                    return

        try:
            if osPathExists(path):
                with open(path, "r") as f:
                    self.data["common"] = self.data.get("common", {})
                    self.data["common"]["public_key"] = f.read().strip()
                    self.logger.info(f"{self.Name} : Public Key Loaded from {path}")
        except Exception as e:
            self.logger.warning(f"{self.Name} : Failed to load public key from {path}: {e}")

    @property
    def common(self) -> Dict[str, Any]:
        """Provides direct access to the 'common' configuration block."""
        return self.data.get("common", {})

    # -----------------------------------------------------------------------------------------------

    def decrypt_secret(self, ciphertext: str) -> str:
        """
        Explicitly decrypts a single ENC(...) ciphertext string.
        Uses the hardened distributed-config engine to ensure cross-language consistency.
        Raises ValueError if decryption fails for an ENC(...) block.
        """
        if not isinstance(ciphertext, str) or not ciphertext.startswith("ENC(") or not ciphertext.endswith(")"):
            return ciphertext

        if self._handle:
            try:
                res = lib.DistConf_Decrypt(self._handle, ciphertext.encode('utf-8'))
                if res:
                    val = res.decode('utf-8')
                    return val
                else:
                    # Fetch raw error from bridge
                    err = lib.DistConf_GetLastError()
                    msg = err.decode('utf-8') if err else "Unknown decryption error"
                    raise ValueError(msg)
            except Exception as e:
                if isinstance(e, ValueError):
                    raise
                self.logger.error(f"{self.Name} : Decryption failed via bridge: {e}")

        # If we reach here, it's an ENC(...) block but we couldn't decrypt it
        # (either bridge failed or bridge is missing). Original behavior was to raise ValueError.
        raise ValueError(f"{self.Name} : Decryption not available or failed for ENC block")

    # -----------------------------------------------------------------------------------------------

    def _load_from_file(self, filename: str) -> None:
        """Full merge of all file data into self.data"""
        file_data = self._read_and_expand_yaml(filename)
        if file_data:
            self.deep_merge(self.data, file_data)

    # -----------------------------------------------------------------------------------------------

    def _apply_file_override(self, filename: str) -> None:
        """Re-reads file and merges ONLY capabilities as hard override (matches Go applyFileOverride)"""
        file_data = self._read_and_expand_yaml(filename)
        if file_data:
            if "capabilities" in file_data:
                self.data["capabilities"] = self.data.get("capabilities", {})
                self.deep_merge(self.data["capabilities"], file_data["capabilities"])
            if "local" in file_data:
                self.data["local"] = self.data.get("local", {})
                self.deep_merge(self.data["local"], file_data["local"])

    def _read_and_expand_yaml(self, filename: str) -> Dict[str, Any]:
        """Helper to read YAML file with environment variable expansion."""
        if not osPathExists(filename):
            return {}

        try:
            with open(filename, "r") as f:
                raw_content = f.read()

            # Expand Environment Variables: ${VAR} or ${VAR:default}
            import re
            def env_expander(match):
                token = match.group(1)
                parts = token.split(":", 1)
                var_name = parts[0]
                default_val = parts[1] if len(parts) > 1 else ""
                return osGetenv(var_name, default_val)

            expanded_content = re.sub(r"\${([^}]+)}", env_expander, raw_content)
            return yamlSafe_load(expanded_content) or {}
        except Exception as e:
            self.logger.warning(f"{self.Name} : Failed to load {filename}: {e}")
            return {}

    # -----------------------------------------------------------------------------------------------

    def _apply_cli_overrides(self) -> None:
        if self.cli_args.name:
            self.data["common"] = self.data.get("common", {})
            self.data["common"]["name"] = self.cli_args.name

        # If network flags provided and not blocked by Docker Guard
        if any([self.cli_args.host, self.cli_args.port, self.cli_args.grpc_host, self.cli_args.grpc_port]):
            target = self.cli_args.name or self.data.get("common", {}).get("name") or "config_server"
            self.data["capabilities"] = self.data.get("capabilities", {})
            cap = self.data["capabilities"].get(target, {})

            if self.cli_args.host:
                cap["ip"] = self.cli_args.host
            if self.cli_args.port:
                cap["port"] = str(self.cli_args.port)
            if self.cli_args.grpc_host:
                cap["grpc_ip"] = self.cli_args.grpc_host
            if self.cli_args.grpc_port:
                cap["grpc_port"] = str(self.cli_args.grpc_port)

            self.data["capabilities"][target] = cap

    # -----------------------------------------------------------------------------------------------

    @staticmethod
    def deep_merge(dst: Dict[str, Any], src: Dict[str, Any]) -> None:
        """Wrapper around standalone deep_merge for backward compatibility."""
        deep_merge(dst, src)

    # -----------------------------------------------------------------------------------------------

    def get_listen_addr(self, capability: str) -> str:
        """
        Resolves the listening address for a capability.
        Prioritizes live bridge resolution (Service Discovery) over local snapshots.
        """
        if self._handle:
            try:
                res = lib.DistConf_GetAddress(self._handle, capability.encode('utf-8'))
                if res:
                    return res.decode('utf-8')
            except Exception as e:
                self.logger.warning(f"{self.Name} : Bridge GetAddress failed: {e}")

        return self._get_addr(capability, "ip", "port")

    # -----------------------------------------------------------------------------------------------

    def get_grpc_listen_addr(self, capability: str) -> str:
        """
        Resolves the gRPC listening address for a capability.
        Requires explicit grpc_ip and grpc_port (matching Go GetGRPCAddress behavior).
        """
        if self._handle:
            try:
                res = lib.DistConf_GetGRPCAddress(self._handle, capability.encode('utf-8'))
                if res:
                    return res.decode('utf-8')
            except Exception as e:
                self.logger.warning(f"{self.Name} : Bridge GetGRPCAddress failed: {e}")

        return self._get_addr(capability, "grpc_ip", "grpc_port")

    # -----------------------------------------------------------------------------------------------

    def _get_addr(self, capability: str, host_key: str, port_key: str) -> str:
        caps = self.data.get("capabilities", {})
        cap = caps.get(capability)
        if not cap:
            raise ValueError(f"capability {capability} not found")

        host = cap.get(host_key)
        if not host:
            raise ValueError(f"host key {host_key} missing or empty in capability {capability}")

        port = cap.get(port_key)
        if not port:
            raise ValueError(f"port key {port_key} missing or empty in capability {capability}")

        return f"{host}:{port}"

    # -----------------------------------------------------------------------------------------------

    def _sync_from_bridge(self) -> None:
        """Pulls the full configuration state from the Go bridge into self.data."""
        if not self._handle:
            return

        try:
            full_json = lib.DistConf_GetFullConfig(self._handle)
            if full_json:
                self.data = jsonLoads(full_json.decode('utf-8'))
        except Exception:
            err = lib.DistConf_GetLastError()
            if err:
                self.logger.warning(err.decode('utf-8'))

    # -----------------------------------------------------------------------------------------------

    def get_config(self, section: str, key: str, default: Any = None) -> Any:
        """Retrieves a value from the bridge or local data."""
        if self._handle:
            res = lib.DistConf_Get(self._handle, section.encode('utf-8'), key.encode('utf-8'))
            if res is not None:
                return res.decode('utf-8')

        # Fallback to local data
        sect = self.data.get(section, {})
        return sect.get(key, default)

    # -----------------------------------------------------------------------------------------------

    def get_local(self, key: str) -> Any:
        """
        Returns a value from the 'local' configuration section.
        Supports nested lookups using dot notation (e.g., "database.host").
        """
        local_data = self.data.get("local")
        if local_data is None:
            return None

        parts = key.split(".")
        current = local_data
        for part in parts:
            if isinstance(current, dict):
                current = current.get(part)
                if current is None:
                    return None
            else:
                return None
        return current

    # -----------------------------------------------------------------------------------------------

    def unmarshal_local(self, target: Any) -> Any:
        """
        Unmarshals the 'local' configuration section into a target type or instance.
        Parity with Go's UnmarshalLocal.
        """
        local_data = self.data.get("local")
        if not local_data:
            raise ValueError("No local configuration found")

        import json
        raw_json = json.dumps(local_data)
        data = json.loads(raw_json)

        if isinstance(target, type):
            # If target is a class (e.g. a dataclass), instantiate it
            return target(**data)
        else:
            # If target is an existing instance, update its attributes
            for k, v in data.items():
                setattr(target, k, v)
            return target

    # -----------------------------------------------------------------------------------------------

    def set_logger(self, logger: Any) -> None:
        """Updates the logger after instantiation."""
        self.logger = ensure_safe_logger(logger)
        self.logger.info("{0} : Logger updated successfully".format(self.Name))

    # -----------------------------------------------------------------------------------------------

    def on_live_conf_update(self, callback: Callable[[Dict[str, Any]], None]) -> None:
        """Registers a callback for live configuration updates (Dynamic Fleet Config)."""
        if not self._handle:
            self.logger.warning(f"{self.Name} : Live updates not supported (no bridge)")
            return

        def _bridge_cb(handle: int, json_data: bytes) -> None:
            try:
                data = jsonLoads(json_data.decode('utf-8'))
                callback(data)
                self._sync_from_bridge() # Keep local data in sync
            except Exception as e:
                self.logger.error(f"{self.Name} : Live update callback failed: {e}")

        self._callback_refs["live"] = CALLBACK_TYPE(_bridge_cb)
        lib.DistConf_OnLiveConfUpdate(self._handle, self._callback_refs["live"])

    # -----------------------------------------------------------------------------------------------

    def on_registry_update(self, callback: Callable[[Dict[str, Any]], None]) -> None:
        """Registers a callback for service registry changes (Service Discovery)."""
        if not self._handle:
            self.logger.warning(f"{self.Name} : Registry updates not supported (no bridge)")
            return

        def _bridge_cb(handle: int, json_data: bytes) -> None:
            try:
                data = jsonLoads(json_data.decode('utf-8'))
                callback(data)
                self._sync_from_bridge()
            except Exception as e:
                self.logger.error(f"{self.Name} : Registry update callback failed: {e}")

        self._callback_refs["registry"] = CALLBACK_TYPE(_bridge_cb)
        lib.DistConf_OnRegistryUpdate(self._handle, self._callback_refs["registry"])

    # -----------------------------------------------------------------------------------------------

    def on_update(self, callback: Callable[[Dict[str, Any]], None]) -> None:
        """Deprecated: Use on_live_conf_update(). Semantic alias for backward compatibility."""
        self.on_live_conf_update(callback)

    # -----------------------------------------------------------------------------------------------

    def share_config(self, payload: Dict[str, Any]) -> bool:
        """Shares service configuration with the ecosystem."""
        if not self._handle:
            return False
        try:
            return lib.DistConf_ShareConfig(self._handle, jsonDumps(payload).encode('utf-8'))
        except Exception as e:
            self.logger.error(f"{self.Name} : ShareConfig failed: {e}")
            return False

    # -----------------------------------------------------------------------------------------------

    def close(self):
        """Releases the underlying DistConf handle."""
        if hasattr(self, "_handle") and self._handle:
            self._lib.DistConf_Close(self._handle)
            self._handle = None

    def __del__(self):
        self.close()

File: go/pkg/config/loader.go
// -----------------------------------------------------------------------------
// ESSENTIAL PROCESS:
// Loads layered configuration from local YAML, environment variables, and distributed config server.
//
// DATA FLOW:
// YAML File / Env Vars / Remote Server -> Config Parsing -> AppConfig Instance
//
// KEY PARAMETERS:
// - profile: Configuration profile (standalone, test, staging, production).
// - data: In-memory configuration key-value tree.
// -----------------------------------------------------------------------------

package config

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"github.com/Bastien-Antigravity/microservice-toolbox/go/pkg/connectivity"
	"github.com/Bastien-Antigravity/microservice-toolbox/go/pkg/logger"

	distributed_config "github.com/Bastien-Antigravity/distributed-config"

	"gopkg.in/yaml.v3"
)

// AppConfig wraps the distributed-config and provides toolbox enhancements.
type AppConfig struct {
	*distributed_config.Config
	Local    map[string]interface{}
	Resolver *connectivity.Resolver
	Profile  string
	Logger   logger.Logger
	Args     *CLIArgs
}

// GetServiceName returns the standardized program name.
func (ac *AppConfig) GetServiceName() string {
	if ac.Common.Name != "" {
		return ac.Common.Name
	}
	return "unknown-service"
}

// SetLogger updates the logger after instantiation.
func (ac *AppConfig) SetLogger(l logger.Logger) {
	ac.Logger = logger.EnsureSafeLogger(l)
	ac.Logger.Info("Logger updated successfully")
}

// LoadConfig loads the configuration with layered priority:
// 1. CLI Flags (Highest)
// 2. Layered Merge (File/Server based on Profile)
// 3. Env Vars (Lowest)
// LoadConfig loads the configuration with layered priority.
func LoadConfig(profile string, specificFlags []string) (*AppConfig, error) {
	return LoadConfigWithLogger(profile, nil, specificFlags)
}

// LoadConfigWithLogger loads the configuration with an explicit logger and layered priority.
func LoadConfigWithLogger(profile string, l logger.Logger, specificFlags []string) (*AppConfig, error) {
	safeLogger := logger.EnsureSafeLogger(l)

	// Step 0: Preliminary CLI Parse to handle --profile override
	tempAC := &AppConfig{
		Logger:   safeLogger,
		Resolver: connectivity.NewResolver(),
	}
	cliArgs := tempAC.ParseCLIArgs(specificFlags)

	actualProfile := profile
	if cliArgs.Profile != "" {
		actualProfile = cliArgs.Profile
	}

	safeLogger.Info("Initializing Config with Profile: %s", actualProfile)

	// Phase 1: Base Configuration (Initialize Distributed Config)
	// This loads the base YAML and performs initial Remote Sync if necessary.
	dConf := distributed_config.New(actualProfile)
	if dConf == nil {
		return nil, fmt.Errorf("failed to load distributed config for profile: %s", actualProfile)
	}

	ac := &AppConfig{
		Config:   dConf,
		Resolver: connectivity.NewResolver(),
		Profile:  actualProfile,
		Logger:   safeLogger,
		Args:     cliArgs,
	}

	// Phase 2: Local Overrides (Developer-Intent Parity)
	// We re-apply the local file as a hard override to ensure the 'local' section
	// and any local overrides are loaded across all profiles.
	ac.Logger.Info("Applying Local File as Hard Override (Ecosystem Parity).")
	ac.applyFileOverride(actualProfile + ".yaml")

	// Phase 3: CLI Arguments (Highest Priority)
	ac.applyCLIOverrides(cliArgs)
	ac.applyCLIGRPCOverrides(cliArgs)

	// Inject the manually defined local config from the test if it was set before LoadConfig (though unlikely in prod)
	// but mostly to ensure we don't wipe it out if it was already there.
	if tempAC.Local != nil {
		ac.Local = DeepMerge(ac.Local, tempAC.Local)
	}

	// Ensure the service name is synchronized to the base common config
	if cliArgs.Name != "" {
		ac.Common.Name = cliArgs.Name
	}

	// Phase 4: Remote Sync (Managed by the base Distributed Config)
	// The sync happened in Phase 1, but we ensure consistency here.

	// Phase 5: Security & Keys
	ac.loadPublicKey()

	// If --key flag provided, set it as ENV override for the Private Key (decryption engine)
	if cliArgs.Key != "" {
		_ = os.Setenv("BASTIEN_PRIVATE_KEY_PATH", cliArgs.Key)
	}

	if err := ac.ValidateUniquePorts(); err != nil {
		return nil, fmt.Errorf("configuration validation failed: %w", err)
	}

	return ac, nil
}

func (ac *AppConfig) loadPublicKey() {
	path := os.Getenv("BASTIEN_PUBLIC_KEY_PATH")
	if path == "" {
		candidates := []string{"/etc/bastien/public.pem", "./public.pem"}
		for _, c := range candidates {
			if _, err := os.Stat(c); err == nil {
				path = c
				break
			}
		}
	}

	if path != "" {
		if content, err := os.ReadFile(path); err == nil {
			ac.Config.Common.PublicKey = string(content)
		}
	}
}

func (ac *AppConfig) applyFileOverride(filename string) {
	// Build search candidates (matches distributed-config logic)
	targetName := strings.TrimSuffix(filename, filepath.Ext(filename))
	path := distributed_config.ResolveConfigPath(targetName)

	data, err := os.ReadFile(path)
	if err != nil {
		return
	}

	var root yaml.Node
	if err := yaml.Unmarshal(data, &root); err != nil {
		return
	}

	// Expand Environment Variables and force types using Distributed Config's logic
	distributed_config.ProcessNode(&root)

	// Decode into a generic map to extract sections
	var raw map[string]interface{}
	if err := root.Decode(&raw); err != nil {
		return
	}

	// Extract sections
	if caps, ok := raw["capabilities"].(map[string]interface{}); ok {
		ac.Config.Capabilities = DeepMerge(ac.Config.Capabilities, caps)
	}
	if comm, ok := raw["common"].(map[string]interface{}); ok {
		if name, ok := comm["name"].(string); ok {
			ac.Config.Common.Name = name
		}
	}
	if priv, ok := raw["local"].(map[string]interface{}); ok {
		if ac.Local == nil {
			ac.Local = make(map[string]interface{})
		}
		ac.Local = DeepMerge(ac.Local, priv)
	}
}

// GetLocal returns a value from the 'local' configuration section.
// Supports nested lookups using dot notation (e.g., "database.host").
func (ac *AppConfig) GetLocal(key string) interface{} {
	if ac == nil || ac.Local == nil {
		return nil
	}

	parts := strings.Split(key, ".")
	var current interface{} = ac.Local

	for _, part := range parts {
		if m, ok := current.(map[string]interface{}); ok {
			if val, exists := m[part]; exists {
				current = val
			} else {
				return nil
			}
		} else {
			return nil
		}
	}

	return current
}

// UnmarshalLocal maps the 'local' configuration section into a target struct.
func (ac *AppConfig) UnmarshalLocal(target interface{}) error {
	if ac.Local == nil {
		return fmt.Errorf("no local configuration found")
	}
	data, err := yaml.Marshal(ac.Local)
	if err != nil {
		return err
	}
	return yaml.Unmarshal(data, target)
}

// DecryptSecret decrypts a single ENC(...) ciphertext string.
// If the string does not start with ENC(...), it is returned as-is (plaintext fallback).
// If it is an ENC(...) block but decryption fails, an error is returned.
func (ac *AppConfig) DecryptSecret(ciphertext string) (string, error) {
	if !strings.HasPrefix(ciphertext, "ENC(") || !strings.HasSuffix(ciphertext, ")") {
		return ciphertext, nil
	}
	return distributed_config.Decrypt(ciphertext)
}

// OnLiveConfUpdate registers a callback for live configuration updates.
func (ac *AppConfig) OnLiveConfUpdate(cb func(map[string]interface{})) {
	ac.Config.OnLiveConfUpdate(func(data map[string]map[string]string) {
		// Convert to generic map for the toolbox's uniform API
		generic := make(map[string]interface{})
		for k, v := range data {
			inner := make(map[string]interface{})
			for ik, iv := range v {
				inner[ik] = iv
			}
			generic[k] = inner
		}
		cb(generic)
	})
}

// OnRegistryUpdate registers a callback for service registry changes.
func (ac *AppConfig) OnRegistryUpdate(cb func(map[string]interface{})) {
	ac.Config.OnRegistryUpdate(func(data map[string][]string) {
		generic := make(map[string]interface{})
		for k, v := range data {
			generic[k] = v
		}
		cb(generic)
	})
}

// ShareConfig shares service configuration with the ecosystem.
func (ac *AppConfig) ShareConfig(payload map[string]interface{}) error {
	return ac.Config.ShareConfig(payload)
}

func (ac *AppConfig) applyCLIOverrides(args *CLIArgs) {
	if args.Name != "" {
		ac.Config.Common.Name = args.Name
	}

	target := args.Name
	if target == "" {
		target = ac.Config.Common.Name
	}
	if target == "" {
		target = "config_server"
	}

	if args.Host != "" || args.Port != 0 {
		ac.ensurePath("capabilities." + target)
		cap := ac.Config.Capabilities[target].(map[string]interface{})
		if args.Host != "" {
			cap["ip"] = args.Host
		}
		if args.Port != 0 {
			cap["port"] = fmt.Sprintf("%d", args.Port)
		}
	}
}

func (ac *AppConfig) applyCLIGRPCOverrides(args *CLIArgs) {
	target := args.Name
	if target == "" {
		target = ac.Config.Common.Name
	}
	if target == "" {
		target = "config_server"
	}

	if args.GRPCHost != "" || args.GRPCPort != 0 {
		ac.ensurePath("capabilities." + target)
		cap := ac.Config.Capabilities[target].(map[string]interface{})
		if args.GRPCHost != "" {
			cap["grpc_ip"] = args.GRPCHost
		}
		if args.GRPCPort != 0 {
			cap["grpc_port"] = fmt.Sprintf("%d", args.GRPCPort)
		}
	}
}

func (ac *AppConfig) ensurePath(path string) {
	if ac.Config.Capabilities == nil {
		ac.Config.Capabilities = make(map[string]interface{})
	}
	parts := strings.Split(path, ".")
	if len(parts) < 2 {
		return
	}
	// Currently we only support capabilities.NAME
	if parts[0] == "capabilities" {
		target := parts[1]
		if _, ok := ac.Config.Capabilities[target].(map[string]interface{}); !ok {
			ac.Config.Capabilities[target] = make(map[string]interface{})
		}
	}
}

func (ac *AppConfig) GetListenAddr(capability string) (string, error) {
	return ac.Config.GetAddress(capability)
}

func (ac *AppConfig) GetGRPCListenAddr(capability string) (string, error) {
	return ac.Config.GetGRPCAddress(capability)
}

// GetRESTAddr returns the address for the REST management interface.
func (ac *AppConfig) GetRESTAddr(capability string) (string, error) {
	return ac.Config.GetRESTAddress(capability)
}


type endpointKey struct {
	IP   string
	Port int
}

func normalizeIP(ip string) string {
	ip = strings.TrimSpace(strings.ToLower(ip))
	if ip == "localhost" {
		return "127.0.0.1"
	}
	return ip
}

func getPort(val interface{}) (int, bool) {
	if val == nil {
		return 0, false
	}
	switch v := val.(type) {
	case int:
		return v, true
	case int64:
		return int(v), true
	case float64:
		return int(v), true
	case string:
		p, err := strconv.Atoi(v)
		if err == nil {
			return p, true
		}
	}
	return 0, false
}

func getIP(val interface{}) string {
	if val == nil {
		return ""
	}
	return fmt.Sprintf("%v", val)
}

func (ac *AppConfig) ValidateUniquePorts() error {
	if ac.Config == nil || ac.Config.Capabilities == nil {
		return nil
	}

	seen := make(map[endpointKey]string)

	checkAndAdd := func(ip string, port int, desc string) error {
		nip := normalizeIP(ip)
		for seenKey, seenDesc := range seen {
			if seenKey.Port == port {
				if nip == seenKey.IP || nip == "0.0.0.0" || seenKey.IP == "0.0.0.0" {
					return fmt.Errorf("duplicate endpoint detected: %s:%d for %s conflicts with %s:%d for %s", ip, port, desc, seenKey.IP, seenKey.Port, seenDesc)
				}
			}
		}
		seen[endpointKey{IP: nip, Port: port}] = desc
		return nil
	}

	for capName, capVal := range ac.Config.Capabilities {
		capMap, ok := capVal.(map[string]interface{})
		if !ok {
			continue
		}

		ip := getIP(capMap["ip"])
		gIP := getIP(capMap["grpc_ip"])
		if gIP == "" {
			gIP = ip
		}

		// Explicit TCP port
		if port, ok := getPort(capMap["port"]); ok && ip != "" {
			if err := checkAndAdd(ip, port, fmt.Sprintf("%s (TCP)", capName)); err != nil {
				return err
			}
		}

		// Explicit gRPC port
		if gp, ok := getPort(capMap["grpc_port"]); ok && gIP != "" {
			if err := checkAndAdd(gIP, gp, fmt.Sprintf("%s (gRPC)", capName)); err != nil {
				return err
			}
		}

		// Explicit REST port
		if rp, ok := getPort(capMap["rest_port"]); ok && ip != "" {
			if err := checkAndAdd(ip, rp, fmt.Sprintf("%s (REST)", capName)); err != nil {
				return err
			}
		}
	}
	return nil
}
